### PR TITLE
Sync fix from development branch

### PR DESCRIPTION
**Pull request by tgsong**
*Description of changes:*
This PR syncs various fix from development branch. List of changes:

- use gcc memory and undefined behavior sanitizer
- unit tests
- posix timers
- missing PAL header
- PAL callback initialization
- timer call in OTA suspend
- compiler warnings

~~**Note**: the unit tests are still failing, pending fix for~~

- ~~memory leaks in job parsing~~
- ~~missing `pthread_exit` in ota agent unit tests~~

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: source/ota_mqtt.c [hunk 1]
```diff
@@ -848,7 +821,7 @@ OtaErr_t requestFileBlock_Mqtt( OtaAgentContext_t * pAgentCtx )
     /* Reset number of blocks requested. */
     pAgentCtx->numOfBlocksToReceive = otaconfigMAX_NUM_BLOCKS_REQUEST;
 
-    uint32_t blockSize = OTA_FILE_BLOCK_SIZE & 0x7fffffffU;
+    blockSize = OTA_FILE_BLOCK_SIZE & 0x7fffffffU;
```
Comment (yanjos-dev) [inline source/ota_mqtt.c:824]: where does the constant "0x7fffffffU" come from? Can you make a macro for this for readability or add a comment?

File: test/CMakeLists.txt [hunk 2]
```diff
@@ -92,6 +92,6 @@ add_subdirectory( unit-test )
 add_custom_target( coverage
     COMMAND ${CMAKE_COMMAND} -DCMOCK_DIR=${CMOCK_DIR}
     -P ${MODULE_ROOT_DIR}/tools/cmock/coverage.cmake
-    DEPENDS cmock unity ota_utest ota_base64_utest ota_job_parsing_utest ota_cbor_utest ota_http_utest ota_os_posix_utest
+    DEPENDS cmock unity ota_utest ota_base64_utest ota_job_parsing_utest ota_cbor_utest ota_os_posix_utest
```
Comment (yanjos-dev) [inline test/CMakeLists.txt:95]: Why remove the ota http unit tests?

File: test/unit-test/CMakeLists.txt [hunk 3]
```diff
@@ -109,12 +109,16 @@ create_test(ota_cbor_utest
 create_test(ota_os_posix_utest
     "ota_os_posix_utest.c"
     "${utest_link_list}"
-    "${utest_mem_ops_list}"
```
Comment (yanjos-dev) [inline test/unit-test/CMakeLists.txt:112]: Where was utest_mem_ops _list defined? Does that definition need to be removed if it's not being used anymore?

File: test/unit-test/ota_job_parsing_utest.c [hunk 4]
```diff
@@ -88,33 +70,35 @@ const AppVersion32_t appFirmwareVersion =
 /* OTA code signing signature algorithm. */
 const char OTA_JsonFileSignatureKey[ OTA_FILE_SIG_KEY_STR_MAX_LENGTH ] = "sig-sha256-ecdsa";
 
+/* OTA interfaces. */
+static OtaInterfaces_t otaInterfaces;
+
+/* OTA application buffer. */
+static OtaAppBuffer_t otaAppBuffer;
+static uint8_t pUserBuffer[ 300 ];
```
Comment (yanjos-dev) [inline test/unit-test/ota_job_parsing_utest.c:78]: Where does 300 come from? There should be a macro for this.

File: test/unit-test/ota_job_parsing_utest.c [hunk 5]
```diff
@@ -88,33 +70,35 @@ const AppVersion32_t appFirmwareVersion =
...
-        { OTA_JSON_FILE_SIZE_KEY,       OTA_JOB_PARAM_REQUIRED, { offsetof( OtaFileContext_t, fileSize )}, ModelParamTypeUInt32      },
-        { OTA_JSON_FILE_ID_KEY,         OTA_JOB_PARAM_REQUIRED, { offsetof( OtaFileContext_t, serverFileID )}, ModelParamTypeUInt32      },
-        { OTA_JSON_FILE_CERT_NAME_KEY,  OTA_JOB_PARAM_REQUIRED, { offsetof( OtaFileContext_t, pCertFilepath )}, ModelParamTypeStringCopy  },
-        { OTA_JSON_UPDATE_DATA_URL_KEY, OTA_JOB_PARAM_OPTIONAL, { offsetof( OtaFileContext_t, pUpdateUrlPath )}, ModelParamTypeStringCopy  },
-        { OTA_JSON_AUTH_SCHEME_KEY,     OTA_JOB_PARAM_OPTIONAL, { offsetof( OtaFileContext_t, pAuthScheme )}, ModelParamTypeStringCopy  },
-        { OTA_JsonFileSignatureKey,     OTA_JOB_PARAM_REQUIRED, { offsetof( OtaFileContext_t, pSignature )}, ModelParamTypeSigBase64   },
-        { OTA_JSON_FILE_ATTRIBUTE_KEY,  OTA_JOB_PARAM_OPTIONAL, { offsetof( OtaFileContext_t, fileAttributes )}, ModelParamTypeUInt32      },
-    };
+    otaAppBuffer.pUpdateFilePath = pUserBuffer;
+    otaAppBuffer.updateFilePathsize = 100;
```
Comment (yanjos-dev) [inline test/unit-test/ota_job_parsing_utest.c:85]: Can there be macros for these constants? This way others will know why they are these values and when they need to be updated.

File: test/CMakeLists.txt (hunk 2 above)
Comment (tgsong) [reply test/CMakeLists.txt:95]: It has not been added yet, probably a mistake from previous pr.

File: test/unit-test/CMakeLists.txt (hunk 3 above)
Comment (tgsong) [reply test/unit-test/CMakeLists.txt:112]: It's not defined anywhere, so I'm removing it.

File: source/ota_mqtt.c (hunk 1 above)
Comment (tgsong) [reply source/ota_mqtt.c:824]: I don't know where it came from, I was merely fixing compiler warnings by moving the declaration to top. @pvyawaha any idea of this magic number?

**Review by yanjos-dev — APPROVED** [review]

File: source/ota_mqtt.c (hunk 1 above)
Comment (tgsong) [reply source/ota_mqtt.c]: Looks like it's a legacy workaround, https://github.com/aws/amazon-freertos/blob/be79e3a653bd8acf2ea575662c59e33be0439773/libraries/freertos_plus/aws/ota/src/mqtt/aws_iot_ota_mqtt.c#L848. I'll remove it.

**Review by pvyawaha — APPROVED** [review]

**Review by divekarshubham — APPROVED** [review]

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -50,7 +50,7 @@ jobs:
         uses: ChicagoFlutter/lcov-cop@v1.0.2
         with:
           path: "build/coverage.info"
-          min_coverage: 100
+          min_coverage: 80
```
Comment (dan4thewin) [inline .github/workflows/ci.yml:53]: I trust this is only a temporary change.

File: source/ota.c [hunk 7]
```diff
@@ -1232,7 +1240,7 @@ static DocParseErr_t extractAndStoreArray( char * pKey,
             err = DocParseErrOutOfMemory;
 
             LogError( ( "Memory allocation failed "
-                        "[key: valueLength]=[%s: %s]",
+                        "[key: valueLength]=[%s: %lu]",
                         pKey,
                         valueLength ) );
```
Comment (dan4thewin) [inline source/ota.c:1245]: I think valueLength here should be cast to a long.  We recently fixed logging in the rest of the repos, such that values are cast to standard types, and int32_t is cast to long.  These changes should be made wherever format strings are used.  See https://github.com/FreeRTOS/coreMQTT/pull/113

Of course, it might be better to do such in a separate PR.

**Review by dan4thewin — APPROVED** [review]

File: source/ota.c (hunk 7 above)
Comment (pvyawaha) [reply source/ota.c:1245]: I will make a note of this and raise another PR.
